### PR TITLE
add exponential simplex transform

### PR DESCRIPTION
**Pull request by spinkney**
After some algebra and cancellation this transform is ready. By preliminary wall clock time and ESS this performs really well.

### CODE REVIEW

**Comment by spinkney** [comment]
@bob-carpenter @sethaxen this is updated from the emails I sent out. It seems to work quite well.

File: simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan [hunk 1]
```diff
@@ -0,0 +1,21 @@
+vector exponential_simplex_constrain_lp(vector y) {
+  int N = rows(y);
+  vector[N] x_pos = y - log1p_exp(-y);
+  real r = log_sum_exp(x_pos);
+  vector[N] x = exp(x_pos - r);
+  target += x_pos;
+  target += std_normal_lpdf(r - log(N));
+  target += sum(y) - N * r;
+  return x;
```
Comment (spinkney) [inline simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan:8]: I can just return `exp(x_pos - r)` instead of allocating to `x` to save on memory.

File: simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+vector exponential_log_simplex_constrain_lp(vector y) {
+  int N = rows(y);
+  vector[N] x = y - log1p_exp(-y);
+  real r = log_sum_exp(x);
+  vector[N] log_x = x - r;
+  target += x;
+  target += std_normal_lpdf(r - log(N));
+  target += log_x[N];
+  return log_x;
```
Comment (spinkney) [inline simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan:18]: Same here, I can return `x - r` instead of allocating but will need to make the last term `log_x[N]`

File: simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan [hunk 3]
```diff
@@ -0,0 +1,21 @@
+vector exponential_simplex_constrain_lp(vector y) {
+  int N = rows(y);
+  vector[N] x_pos = y - log1p_exp(-y);
```
Comment (sethaxen) [inline simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan]: For consistency with the other implementations, perhaps use `z` here:
```suggestion
  vector[N] z = y - log1p_exp(-y);
```

File: simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan (hunk 1 above)
Comment (sethaxen) [reply simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan:8]: Either way `x` would be allocated, I think, since that's needed in the model. Besides, our other implementations explicitly define `x`, so I prefer keeping that for consistency.

File: simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan (hunk 2 above)
Comment (sethaxen) [reply simplex_transforms/stan/transforms/Exponential/Exponential_functions.stan:18]: Like above, I prefer to explicitly define `log_x`.

**Review by sethaxen — CHANGES_REQUESTED** [review]
Thanks for adding! Is the derivation written down somewhere? I may have missed it in one of the e-mails, but can you say something about the rationale behind this transform? So far as I can tell, it's very similar to Expanded-Softmax. In particular, if $f$ is the Expanded-Softmax constraining transform, then as $y_i \to \infty$, this becomes $f(y)$, and as $y_i \to -\infty$, it becomes $f(2y)$. Due to metric adaptation, we wouldn't expect $f(2y)$ to perform any differently than $f(y)$, so really it's just in $\sim(-3, 3)$ when it smoothly transitions between the two that we expect a difference.

On that note, is there a more descriptive name than "Exponential"?

Can you also add a jax implementation and add it to the list of transforms in the tests so that the Jacobian determinant is automatically checked?

**Comment by spinkney** [comment]
> Thanks for adding! Is the derivation written down somewhere?

I'll write it out and send it.

> So far as I can tell, it's very similar to Expanded-Softmax. In particular, if f is the Expanded-Softmax constraining transform, then as y i → ∞ , this becomes f ( y ) , and as y i → − ∞ , it becomes f ( 2 y ) . Due to metric adaptation, we wouldn't expect f ( 2 y ) to perform any differently than f ( y ) , so really it's just in ∼ ( − 3 , 3 ) when it smoothly transitions between the two that we expect a difference.

I was surprised at how `exp(y) / (1 + exp(-y))` had higher ess than `exp(y)` for a positive transform. This is probably just in the range of the inits (-2, 2) that you note is likely. I'm just thinking that a slower increasing function may perform better. 

> On that note, is there a more descriptive name than "Exponential"?

It's more like a generalized logit function I guess.

> Can you also add a jax implementation and add it to the list of transforms in the tests so that the Jacobian determinant is automatically checked?

Ooph. Yea I'll have to use an llm to help me with that but I can try!

**Comment by bob-carpenter** [comment]
> exp(y) / (1 + exp(-y)) had higher ess than exp(y) for a positive transform. 

That's interesting.  I found that log(1 + exp(y)) did not perform better than exp(y) for lognormal, gamma, or inverse gamma or weibull distributions.  It did a little bit better for half normal distributions.  We should've started with an analysis of positive transforms.

**Comment by sethaxen** [comment]
> Ooph. Yea I'll have to use an llm to help me with that but I can try!

If it helps, check out the last 3 commits on https://github.com/bob-carpenter/transforms/tree/ilr2. Following the same approach would work here.
